tools/releaseTools.psm1

```diff
@@ -0,0 +1,184 @@
+        ## commits in the master branch that were cherry-picked to the release branch -- they are reachable from 'HEAD'
+        ## but not reachable from the last release tag. Instead, we need to exclude the commits that were cherry-picked,
+        ## and only include the commits that are not in the last release into the change log.
+
+        # Find the merge commit that merged the release branch to master.
+        $child_merge_commit = Get-ChildMergeCommit -CommitHash $tag_hash
+        $commit_hash, $parent_hashes = $child_merge_commit.Split("|")
+        # Find the other parent of the merge commit, which represents the original head of master right before merging.
+        $other_parent_hash = ($parent_hashes -replace $tag_hash).Trim()
+
+        # Find the commits that were only in the orginal master, excluding those that were cherry-picked to release branch.
+        $new_commits_from_other_parent = git --no-pager log --no-merges --cherry-pick --right-only "$tag_hash...$other_parent_hash" --format=$format | New-CommitNode
+        # Find the commits that were only in the release branch, excluding those that were cherry-picked from master branch.
+        $new_commits_from_last_release = git --no-pager log --no-merges --cherry-pick  --left-only "$tag_hash...$other_parent_hash" --format=$format | New-CommitNode
+        # Find the commits that are actually duplicate but having different patch-ids due to resolving conflicts during the cherry-pick.
+        $duplicate_commits = Compare-Object $new_commits_from_last_release $new_commits_from_other_parent -Property PullRequest -ExcludeDifferent -IncludeEqual -PassThru
+        if ($duplicate_commits) {
+            $duplicate_pr_numbers = @($duplicate_commits | ForEach-Object -MemberName PullRequest)
+            $new_commits_from_other_parent = $new_commits_from_other_parent | Where-Object PullRequest -NotIn $duplicate_pr_numbers
+        }
+
+        # Find the commits that were made after the merge commit.
+        $new_commits_after_merge_commit = git --no-pager log --no-merges "$commit_hash..HEAD" --format=$format | New-CommitNode
+        $new_commits = $new_commits_after_merge_commit + $new_commits_from_other_parent
+    } else {
+        ## No cherry-pick was involved in the last release branch.
+        ## We can get all new commits using the revision range "$tag_hash..HEAD", meaning the commits that are
+        ## reachable from 'HEAD' but not reachable from the last release tag.
+        $new_commits = git --no-pager log --no-merges "$tag_hash..HEAD" --format=$format | New-CommitNode
+    }
+
+    foreach ($commit in $new_commits) {
+        if ($commit.AuthorEmail.EndsWith("@microsoft.com")) {
+            $commit.ChangeLogMessage = "- {0}" -f $commit.Subject
+        } else {
+            $uri = "https://api.github.com/repos/PowerShell/PowerShell/commits/$($commit.Hash)"
```

Comment:
This belongs better in this module: https://github.com/PowerShell/PowerShellForGitHub

Reply:
That module is more about general GitHub operations, while this one is specific to powershell. It's based on how we operate the repository, for example
- we use squash merge almost all the time, so we can ignore merge commits when generating changelog.
- our release process, including how we do cherry-picks and merge release branch to master.
- the format of changelog entries, as well as the assumption that `[breaking change]` tag will be put in commit description body by maintainers.

Reply:
You could also optimize this by getting the author email address locally and only going to github if you don't know the login for that email.

Reply:
Good idea. We don't have to query GitHub for every commit from external contributors, as they may be from the same one. Here is an example of output:
```
- Replace Remaining HttpBin.org Tests with WebListener (#5665) (Thanks @markekraus!)
- Combine Web Cmdlet Partial Class Files (#5612) (Thanks @markekraus!)
- Enable conversions from PSMethod to Delegate (#5287) (Thanks @powercode!)
- Fix PSVersion in PSSessionConfiguration tests (#5554) (Thanks @iSazonov!)
- Update target tag for master CI to 6.1.0-preview.1 (#5513)
- Remove AddTypeCommandBase class (#5407) (Thanks @iSazonov!)
```
We can save 2 queries in this example.

Reply:
Added `$community_login_map` to cache the login names we queried.